File: 4Sum/4Sum.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> result;
        int n = nums.size();
        if (n < 4) return result;

        sort(nums.begin(), nums.end());

        for (int i = 0; i < n - 3; ++i) {
            if (i > 0 && nums[i] == nums[i - 1]) continue; // Skip duplicates for the first number
            for (int j = i + 1; j < n - 2; ++j) {
                if (j > i + 1 && nums[j] == nums[j - 1]) continue; // Skip duplicates for the second number
                int left = j + 1, right = n - 1;
                while (left < right) {
                    long long sum = static_cast<long long>(nums[i]) + nums[j] + nums[left] + nums[right];
                    if (sum == target) {
                        result.push_back({nums[i], nums[j], nums[left], nums[right]});
                        while (left < right && nums[left] == nums[left + 1]) ++left; // Skip duplicates for the third number
                        while (left < right && nums[right] == nums[right - 1]) --right; // Skip duplicates for the fourth number
                        ++left;
                        --right;
                    } else if (sum < target) {
                        ++left;
                    } else {
                        --right;
                    }
                }
            }
        }
        return result;
    }
};
int main() {
    Solution solution;
    vector<int> nums = {1, 0, -1, 0, -2, 2};
    int target = 0;
    vector<vector<int>> result = solution.fourSum(nums, target);
    
    for (const auto& quadruplet : result) {
        cout << "[";
        for (int num : quadruplet) {
            cout << num << " ";
        }
        cout << "]" << endl;
    }
    
    return 0;
}

File: 4Sum/4Sum.rs

fn four_sum(mut nums: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
  let mut result = Vec::new();
  let n = nums.len();
  if n < 4 {
    return result;
  }
  nums.sort();
  for i in 0..n - 3 {
    if i > 0 && nums[i] == nums[i - 1] {
      continue;
    }
    for j in i + 1..n - 2 {
      if j > i + 1 && nums[j] == nums[j - 1] {
        continue;
      }
      let mut left = j + 1;
      let mut right = n - 1;
      while left < right {
        let sum = nums[i] + nums[j] + nums[left] + nums[right];
        if sum == target {
          result.push(vec![nums[i], nums[j], nums[left], nums[right]]);
          while left < right && nums[left] == nums[left + 1] {
            left += 1;
          }
          while left < right && nums[right] == nums[right - 1] {
            right -= 1;
          }
          left += 1;
          right -= 1;
        } else if sum < target {
          left += 1;
        } else {
          right -= 1;
        }
      }
    }
  }
  result
}
fn main() {
  let nums = vec![1, 0, -1, 0, -2, 2];
  let target = 0;
  let result = four_sum(nums, target);
  for r in result {
    println!("{:?}", r);
  }
}